File: generator.cs
using System;
using System.IO;

namespace WaveformGenerator
{
    public enum WaveformType
    {
        Sine,
        Square,
        Sawtooth,
        Triangle
    }

    public class Waveform
    {
        private const int PLAY_SAMPLING_RATE = 8000;
        private const int PRESS_SIGNAL_VOLTAGE = 60;

        public static void GeneratePressWaveform(double frequency, WaveformType waveformType, ref double[] press_waveform)
        {
            int press_waveform_size = (int)(PLAY_SAMPLING_RATE / frequency);
            double angleIncrement = 2 * Math.PI * frequency / PLAY_SAMPLING_RATE;

            for (int i = 0; i < press_waveform_size; i++)
            {
                double angle = i * angleIncrement;
                double normalizedValue;

                switch (waveformType)
                {
                    case WaveformType.Sine:
                        normalizedValue = Math.Sin(angle);
                        break;
                    case WaveformType.Square:
                        normalizedValue = angle % (2 * Math.PI) < Math.PI ? 1 : -1;
                        break;
                    case WaveformType.Sawtooth:
                        normalizedValue = (angle % (2 * Math.PI)) / (2 * Math.PI) * 2 - 1;
                        break;
                    case WaveformType.Triangle:
                        normalizedValue = 4 * Math.Abs(angle % (2 * Math.PI) / (2 * Math.PI) - 0.5) - 1;
                        break;
                    default:
                        throw new ArgumentException("Invalid waveform type");
                }

                press_waveform[i] = UtilsVolt2Amplitude(PRESS_SIGNAL_VOLTAGE / 2 * (normalizedValue + 1));
            }
        }

        // You will need to define the UtilsVolt2Amplitude function here.
        public static short UtilsVolt2Amplitude(double volt)
        {
            short amplitude = (short)(volt * 2047 / 3.6 / 31);
            return amplitude;
        }
    
        class Program
        {
            static void Main(string[] args)
            {
                double frequency = 50;
                WaveformType waveformType = WaveformType.Sine;
                double[] press_waveform = new double[256];

                Waveform.GeneratePressWaveform(frequency, waveformType, ref press_waveform);



                int graphWidth = press_waveform.Length;

                // Create a list of strings representing the waveform data
                List<string> waveformLines = new List<string>();

                for (int i = 0; i < graphWidth; i++)
                {
                    Console.WriteLine($"Sample {i}: {press_waveform[i]}");
                    waveformLines.Add($"{press_waveform[i]}");
                }

                // Save the waveform data to a text file
                File.WriteAllLines("waveform.txt", waveformLines);

            }
        }
    }
}

File: generator.cpp
#include <stdio.h>
#include <math.h>

#define PI 3.14159265
#define PLAY_SAMPLING_RATE 8000
#define PRESS_SIGNAL_VOLTAGE 60
#define WAVEFORM_SIZE 256

typedef enum {
    SINE,
    SQUARE,
    SAWTOOTH,
    TRIANGLE
} WaveformType;

double press_waveform[WAVEFORM_SIZE];

double UtilsVolt2Amplitude(double volt) {
    double amplitude = (volt * 2047 / 3.6 / 31);
    return amplitude;
}

void GeneratePressWaveform(double frequency, WaveformType waveformType) {
    int press_waveform_size = (int)(PLAY_SAMPLING_RATE / frequency);
    double angleIncrement = 2 * PI * frequency / PLAY_SAMPLING_RATE;

    for (int i = 0; i < press_waveform_size; i++) {
        double angle = i * angleIncrement;
        double normalizedValue;

        switch (waveformType) {
            case SINE:
                normalizedValue = sin(angle);
                break;
            case SQUARE:
                normalizedValue = fmod(angle, 2 * PI) < PI ? 1 : -1;
                break;
            case SAWTOOTH:
                normalizedValue = fmod(angle, 2 * PI) / (2 * PI) * 2 - 1;
                break;
            case TRIANGLE:
                normalizedValue = 4 * fabs(fmod(angle, 2 * PI) / (2 * PI) - 0.5) - 1;
                break;
            default:
                normalizedValue = 0;
                break;
        }

        press_waveform[i] = UtilsVolt2Amplitude(PRESS_SIGNAL_VOLTAGE / 2 * (normalizedValue + 1));
    }
}


int main() {
    double frequency = 50;
    WaveformType waveformType = SINE;  // Change this to SQUARE, SAWTOOTH, or TRIANGLE to generate different waveforms
    GeneratePressWaveform(frequency, waveformType);

    FILE *file = fopen("waveform.txt", "w");
    if (file == NULL) {
        printf("Error opening file!\n");
        return 1;
    }

    for (int i = 0; i < WAVEFORM_SIZE; i++) {
        printf("Sample %d: %f\n", i, press_waveform[i]);
        // fprintf(file, "%f\n", press_waveform[i]);
    }

    fclose(file);
    return 0;
}
